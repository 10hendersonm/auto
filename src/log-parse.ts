import { ICommit } from 'parse-git';
import { URL } from 'url';
import join from 'url-join';
import { ILogger } from './GithubRelease';

interface ICommitAuthor {
  name: string;
  email?: string;
  username?: string;
}

interface IGenerateReleaseNotesOptions {
  owner: string;
  repo: string;
  baseUrl: string;
  jira?: string;
  logger: ILogger;
  changelogTitles: { [label: string]: string };
}

export type IExtendedCommit = ICommit & {
  authors: ICommitAuthor[];
  pullRequest?: {
    number: string;
    base?: string;
  };
  jira?: {
    number: string[];
  };
  labels: string[];
  packages?: string[];
};

type CommitParseFunction = (
  commit: IExtendedCommit
) => IExtendedCommit | undefined;

export const createUserLink = (
  author: ICommitAuthor,
  commit: IExtendedCommit,
  options: IGenerateReleaseNotesOptions
) => {
  const githubUrl = new URL(options.baseUrl).origin;

  if (author.username === 'invalid-email-address') {
    return;
  }

  return author.username
    ? `[@${author.username}](${join(githubUrl, author.username)})`
    : author.email || commit.author.email;
};

const createUserLinkList = (
  commit: IExtendedCommit,
  options: IGenerateReleaseNotesOptions
) => {
  const result = new Set<string>();

  commit.authors.forEach(author => {
    const link = createUserLink(author, commit, options);

    if (link) {
      result.add(link);
    }
  });

  return [...result].join(' ');
};

export function filterServiceAccounts(
  commit: IExtendedCommit
): IExtendedCommit | undefined {
  const SERVICE_ACCOUNTS = ['pdbf'];

  if (SERVICE_ACCOUNTS.includes(commit.author.name)) {
    return;
  }

  return commit;
}

export function parsePR(commit: IExtendedCommit): IExtendedCommit {
  const merge = /Merge pull request #(\d+) from (.+)\n([\S\s]+)/;
  const prMatch = commit.comment.match(merge);

  if (!prMatch) {
    return commit;
  }

  return {
    ...commit,
    pullRequest: {
      number: prMatch[1],
      base: prMatch[2]
    },
    comment: prMatch[3]
  };
}

export function parseSquashPR(
  commit: IExtendedCommit
): IExtendedCommit | undefined {
  const firstLine = commit.comment.split('\n')[0];
  const squashMerge = /\(#(\d+)\)$/;

  const squashMergeMatch = firstLine.match(squashMerge);

  if (!squashMergeMatch) {
    return commit;
  }

  return {
    ...commit,
    pullRequest: {
      number: squashMergeMatch[1]
    },
    comment: firstLine
      .substr(0, firstLine.length - squashMergeMatch[0].length)
      .trim()
  };
}

export function parseJira(commit: IExtendedCommit): IExtendedCommit {
  // Support 'JIRA-XXX:' and '[JIRA-XXX]' and '[JIRA-XXX] - '
  const jira = /^\[?([\w]{3,}-\d+)\]?:?\s?[-\s]*([\S ]+)?/;
  const matches = [];

  let currentMatch = commit.comment.match(jira);

  while (currentMatch) {
    matches.push(currentMatch);
    currentMatch = currentMatch[2].match(jira);
  }

  if (!matches.length) {
    return commit;
  }

  return {
    ...commit,
    jira: {
      number: matches.map(match => match[1])
    },
    comment: matches[matches.length - 1][2]
  };
}

function normalizeCommit(commit: ICommit): IExtendedCommit | undefined {
  const parsers: CommitParseFunction[] = [
    filterServiceAccounts,
    parsePR,
    parseSquashPR,
    parseJira
  ];

  const extended: IExtendedCommit = {
    labels: [],
    ...commit,
    authors: [commit.author]
  };

  return parsers.reduce(
    (prev: IExtendedCommit | undefined, parser) =>
      prev === undefined ? undefined : parser(prev),
    extended
  );
}

export function normalizeCommits(commits: ICommit[]): IExtendedCommit[] {
  return commits.map(normalizeCommit).filter(Boolean) as IExtendedCommit[];
}

function generateCommitNote(
  commit: IExtendedCommit,
  options: IGenerateReleaseNotesOptions
): string | undefined {
  let jira = '';
  let pr = '';

  if (commit.jira && options.jira) {
    const link = options.jira ? join(options.jira, ...commit.jira.number) : '';
    jira = `[${commit.jira.number}](${link}): `;
  }

  if (commit.pullRequest) {
    const prLink = join(options.baseUrl, 'pull', commit.pullRequest.number);
    pr = `[#${commit.pullRequest.number}](${prLink})`;
  }

  const user = createUserLinkList(commit, options);

  return `- ${jira}${commit.comment} ${pr}${user ? ` (${user})` : ''}`;
}

const filterLabel = (commits: IExtendedCommit[], label: string) =>
  commits.filter(commit => commit.labels.includes(label));

/**
 * Split commits into changelogTitle sections
 */
function splitCommits(
  commits: IExtendedCommit[],
  changelogTitles: { [label: string]: string }
): { [key: string]: IExtendedCommit[] } {
  let currentCommits = [...commits];

  commits
    .filter(
      commit =>
        (commit.pullRequest || commit.jira) && commit.labels.length === 0
    )
    .map(commit => commit.labels.push('patch'));

  return Object.assign(
    {},
    ...Object.keys(changelogTitles).map(label => {
      const matchedCommits = filterLabel(currentCommits, label);

      if (matchedCommits.length === 0) {
        return {};
      }

      currentCommits = currentCommits.filter(
        commit => !matchedCommits.includes(commit)
      );

      return {
        [label]: matchedCommits
      };
    })
  );
}

interface INotePartition {
  [key: string]: string[];
}

/**
 * Attempt to create a map of monorepo packages
 */
function partitionPackages(
  labelCommits: IExtendedCommit[],
  lineRender: (commit: IExtendedCommit) => string | undefined
) {
  const packageCommits: INotePartition = {};

  labelCommits.map(commit => {
    const line = lineRender(commit);

    if (line) {
      const packages =
        commit.packages && commit.packages.length
          ? commit.packages.map(p => `\`${p}\``).join(', ')
          : 'monorepo';

      if (!packageCommits[packages]) {
        packageCommits[packages] = [];
      }

      packageCommits[packages].push(line);
    }
  });

  return packageCommits;
}

function createLabelSection(
  split: { [key: string]: IExtendedCommit[] },
  options: IGenerateReleaseNotesOptions,
  sections: string[]
) {
  Object.entries(split).forEach(([label, labelCommits]) => {
    const packageCommits = partitionPackages(labelCommits, commit =>
      generateCommitNote(commit, options)
    );

    const pkgCount = Object.keys(packageCommits).length;
    const title = `#### ${options.changelogTitles[label]}`;
    const hasRepoCommits =
      packageCommits.monorepo && packageCommits.monorepo.length > 0;

    if (pkgCount > 0 && (pkgCount !== 1 || !packageCommits.monorepo)) {
      const section = [title];

      if (hasRepoCommits) {
        packageCommits.monorepo.forEach(note => section.push(note));
        delete packageCommits.monorepo;
      }

      Object.entries(packageCommits).map(([pkg, commits]) => {
        section.push(`- ${pkg}`);
        commits.map(note => section.push(`  ${note}`));
      });

      sections.push(section.join('\n'));
    } else if (hasRepoCommits) {
      sections.push([title, ...packageCommits.monorepo].join('\n'));
    }
  });
}

function createAuthorSection(
  split: { [key: string]: IExtendedCommit[] },
  options: IGenerateReleaseNotesOptions,
  sections: string[]
) {
  const authors = Object.values(split)
    .reduce(
      (
        labeledCommits: IExtendedCommit[],
        sectionCommits: IExtendedCommit[]
      ) => [...labeledCommits, ...sectionCommits],
      []
    )
    .reduce((commitAuthors: string[], commit) => {
      commit.authors.map(author => {
        if (author.username === 'invalid-email-address') {
          return commitAuthors;
        }

        const user = createUserLink(author, commit, options);
        const authorString = `- ${author.name} (${user})`;

        if (!commitAuthors.includes(authorString)) {
          commitAuthors.push(authorString);
        }
      });

      return commitAuthors;
    }, []);

  if (authors.length > 0) {
    let authorSection = `#### Authors: ${authors.length}\n`;
    authorSection += authors.join('\n');
    sections.push(authorSection);
  }
}

export default function generateReleaseNotes(
  commits: IExtendedCommit[],
  options: IGenerateReleaseNotesOptions
): string {
  if (commits.length === 0) {
    return '';
  }

  options.logger.verbose.info('Generating release notes for:\n', commits);

  const split = splitCommits(commits, options.changelogTitles);

  options.logger.verbose.info('Split commits into groups');
  options.logger.veryVerbose.info('\n', split);

  const sections: string[] = [];

  createLabelSection(split, options, sections);

  options.logger.verbose.info('Added groups to changelog');

  createAuthorSection(split, options, sections);

  options.logger.verbose.info('Added authors to changelog');

  const result = sections.join('\n\n');

  options.logger.verbose.info('Successfully generated release notes.');

  return result;
}
