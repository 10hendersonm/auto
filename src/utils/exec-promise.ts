import * as childProcess from 'child_process';
import { promisify } from 'util';

const exec = promisify(childProcess.exec);

export default async function execPromise(cmd: string) {
  const { stdout, stderr } = await exec(cmd, { maxBuffer: 1024 * 1000 });

  if (stderr) {
    console.log(stderr);
  }

  return stdout.trim();
}
