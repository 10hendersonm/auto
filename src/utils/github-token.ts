import fs from 'fs';
import path from 'path';
import registryUrl from 'registry-url';
import { promisify } from 'util';
import settingsUrl from './settings-url';

const readFile = promisify(fs.readFile);
const registry = registryUrl();
// tslint:disable-next-line
const normalizedRegistry = registry.replace('http:', '').replace('https:', '');

export default async function getGitHubToken(
  apiUrl = 'https://api.github.com'
): Promise<string> {
  if (process.env.GH_TOKEN) {
    return process.env.GH_TOKEN;
  }

  const helpText = `Try setting an access token up ${settingsUrl(apiUrl)}`;

  if (!process.env.HOME) {
    throw new Error(`Can't find the GH_TOKEN and no HOME defined. ${helpText}`);
  }

  const rcLocation = path.resolve(process.env.HOME, '.npmrc');

  if (fs.existsSync(rcLocation)) {
    const rc = await readFile(rcLocation, 'utf8');

    const regex = new RegExp(`${normalizedRegistry}:_authToken=\(\\w+\)`);
    const token = rc.match(regex);
    if (token && token[1]) {
      return token[1];
    }

    throw new Error(`No token in the .npmrc. ${helpText}`);
  }

  throw new Error(`Can't find a GitHub token to use. ${helpText}`);
}
