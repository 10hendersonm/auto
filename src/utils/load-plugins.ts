import * as path from 'path';
import { AutoRelease } from '../main';
import NPMPlugin from '../plugins/npm';

export type IPluginConstructor = new () => IPlugin;

export interface IPlugin {
  name: string;
  apply(auto: AutoRelease): void;
}

type SupportedPlugin = 'npm';
const plugins = new Map<SupportedPlugin, IPluginConstructor>([
  ['npm', NPMPlugin]
]);

function isSupported(key: SupportedPlugin | string): key is SupportedPlugin {
  return !!plugins.get(key as SupportedPlugin);
}

export default function loadPlugin(
  pluginPath: SupportedPlugin | string
): IPluginConstructor | undefined {
  if (isSupported(pluginPath)) {
    return plugins.get(pluginPath);
  }

  try {
    return require(pluginPath);
  } catch (error) {
    return require(path.join(process.cwd(), pluginPath));
  }
}
