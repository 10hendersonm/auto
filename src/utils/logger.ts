import signale, { Signale } from 'signale';

export interface ILogger {
  log: signale.Signale<signale.DefaultMethods>;
  verbose: signale.Signale<signale.DefaultMethods>;
  veryVerbose: signale.Signale<signale.DefaultMethods>;
}

export function dummyLog(): ILogger {
  return {
    log: new Signale({ disabled: true }),
    verbose: new Signale({ disabled: true }),
    veryVerbose: new Signale({ disabled: true })
  };
}

export type LogLevel = 'verbose' | 'veryVerbose' | undefined;

export default function createLog(mode: LogLevel): ILogger {
  return {
    log: new Signale(),
    verbose: new Signale({
      disabled: mode !== 'verbose' && mode !== 'veryVerbose'
    }),
    veryVerbose: new Signale({
      disabled: mode !== 'veryVerbose'
    })
  };
}
