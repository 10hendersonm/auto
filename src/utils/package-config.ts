import fs from 'fs';
import parseGithubUrl from 'parse-github-url';
import { promisify } from 'util';

export interface IRepoConfig {
  owner: string;
  repo: string;
  version: string;
}

interface IPackageConfig {
  version: string;
  repository?:
    | {
        url: string;
      }
    | 'string';
}

const readFile = promisify(fs.readFile);

const getPackageConfig = async (): Promise<IPackageConfig> => {
  const pkgConfig = await readFile('./package.json', 'utf-8');
  return JSON.parse(pkgConfig);
};

export default async function getConfigFromPackageJson(): Promise<IRepoConfig> {
  const { repository, version } = await getPackageConfig();

  if (!repository) {
    throw new Error('Cannot read repo info from package.json');
  }

  const { owner, name } =
    parseGithubUrl(
      typeof repository === 'string' ? repository : repository.url
    ) || ({} as any);

  if (!owner || !name) {
    throw new Error(
      'Cannot read owner and package name from Github URL in package.json'
    );
  }

  return {
    repo: name,
    owner,
    version
  };
}
