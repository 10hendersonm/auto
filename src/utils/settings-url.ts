import * as url from 'url';

export default function settingsUrl(apiUrl: string) {
  let { host } = url.parse(apiUrl);

  if (host && host.includes('api.')) {
    host = host.replace('api.', '');
  }

  return `https://${host}/settings/tokens/new`;
}
