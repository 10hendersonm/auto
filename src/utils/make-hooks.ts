import { AsyncSeriesBailHook, AsyncSeriesHook, SyncHook } from 'tapable';
import { IGitHubReleaseHooks } from '../github-release';
import { IChangelogHooks } from '../log-parse';
import { IAutoHooks } from '../main';

export const makeHooks = (): IAutoHooks => ({
  beforeRun: new SyncHook(['config']),
  beforeShipIt: new SyncHook([]),
  onCreateGitHubRelease: new SyncHook(['config']),
  onCreateChangelog: new SyncHook(['gitHubReleaseConfig']),
  getAuthor: new AsyncSeriesBailHook([]),
  getPreviousVersion: new AsyncSeriesBailHook(['prefixRelease']),
  getRepository: new AsyncSeriesBailHook([]),
  publish: new AsyncSeriesHook(['version'])
});

export const makeGitHubReleaseHooks = (): IGitHubReleaseHooks => ({
  onCreateChangelog: new SyncHook(['changelog'])
});

export const makeChangelogHooks = (): IChangelogHooks => ({
  renderChangelogLine: new AsyncSeriesBailHook(['commits', 'lineRender']),
  renderChangelogTitle: new AsyncSeriesBailHook(['commits', 'lineRender']),
  renderChangelogAuthor: new AsyncSeriesBailHook([
    'author',
    'commit',
    'options'
  ]),
  renderChangelogAuthorLine: new AsyncSeriesBailHook(['author', 'user'])
});
