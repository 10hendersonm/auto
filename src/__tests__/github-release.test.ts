import GitHubRelease, { VersionLabel } from '../github-release';
import { normalizeCommits } from '../log-parse';
import SEMVER from '../semver';
import { dummyLog } from '../utils/logger';
import makeCommitFromMsg from './make-commit-from-msg';

const logger = dummyLog();

const constructor = jest.fn();
const getGitLog = jest.fn();
const publish = jest.fn();
const getLabels = jest.fn();
const getPullRequest = jest.fn();
const getLatestRelease = jest.fn();
const getSha = jest.fn();
const createStatus = jest.fn();
const getProject = jest.fn();
const createComment = jest.fn();
const changedPackages = jest.fn();
const getCommitsForPR = jest.fn();
const getUserByEmail = jest.fn();
const getUserByUsername = jest.fn();
const getProjectLabels = jest.fn();
const createLabel = jest.fn();
const getPullRequests = jest.fn();

getProject.mockReturnValue({
  data: { html_url: 'https://custom-git.com' }
});

// @ts-ignore
jest.mock('../git.ts', () => (...args) => {
  constructor(...args);
  return {
    options: { owner: 'test', repo: 'test', version: '1.0.0' },
    getGitLog,
    publish,
    getLabels,
    getLatestRelease,
    getPullRequest,
    getSha,
    createStatus,
    createComment,
    getProject,
    changedPackages,
    getCommitsForPR,
    getUserByUsername,
    getUserByEmail,
    getProjectLabels,
    createLabel,
    getPullRequests
  };
});

const execSpy = jest.fn();
// @ts-ignore
jest.mock('../utils/exec-promise.ts', () => (...args) => execSpy(...args));

const existsSync = jest.fn();
const writeSpy = jest.fn();

let readResult = '{}';

jest.mock('fs', () => ({
  // @ts-ignore
  existsSync: (...args) => existsSync(...args),
  // @ts-ignore
  readFile: (a, b, cb) => {
    cb(null, readResult);
  },
  // @ts-ignore
  ReadStream: () => undefined,
  // @ts-ignore
  WriteStream: () => undefined,
  // @ts-ignore
  closeSync: () => undefined,
  // @ts-ignore
  writeFile: (file, data, cb) => {
    cb(null, writeSpy(file, data));
  }
}));

const slackSpy = jest.fn();

jest.mock('../utils/slack.ts', () => () => slackSpy());
jest.mock('../utils/package-config', () => async () => ({}));
jest.mock('../utils/github-token', () => async () => ({}));

describe('GitHubRelease', () => {
  beforeEach(() => {
    getGitLog.mockClear();
    publish.mockClear();
    getLabels.mockClear();
    writeSpy.mockClear();
    execSpy.mockClear();
    execSpy.mockClear();
    createLabel.mockClear();
  });

  test('should use options owner, repo, and token', async () => {
    const gh = new GitHubRelease({
      owner: 'Andrew',
      repo: 'test',
      token: 'MY_TOKENss'
    });

    await gh.getCommits('12345');

    expect(constructor.mock.calls[0][0].owner).toBe('Andrew');
    expect(constructor.mock.calls[0][0].repo).toBe('test');
  });

  describe('getCommits', async () => {
    test('should default to HEAD', async () => {
      const gh = new GitHubRelease();
      await gh.getCommits('12345');
      expect(getGitLog).toHaveBeenCalled();
    });

    test('should use configured HEAD', async () => {
      const gh = new GitHubRelease({ owner: 'Adam Dierkens', repo: 'test' });
      await gh.getCommits('12345', '1234');
      expect(getGitLog).toHaveBeenCalled();
    });

    test('should not resolve authors with no PR commits', async () => {
      const commits = [
        makeCommitFromMsg('First'),
        makeCommitFromMsg('Second'),
        makeCommitFromMsg('Third')
      ];

      getGitLog.mockReturnValueOnce(commits);
      const gh = new GitHubRelease({ owner: 'Adam Dierkens', repo: 'test' });
      await gh.getCommits('12345', '1234');
      expect(getUserByUsername).not.toHaveBeenCalled();
    });

    test('should resolve authors with PR commits', async () => {
      const commits = normalizeCommits([
        makeCommitFromMsg('First'),
        makeCommitFromMsg('Second (#123)', {
          name: 'Andrew Lisowski',
          email: 'lisowski54@gmail.com'
        }),
        makeCommitFromMsg('Third')
      ]);

      getGitLog.mockReturnValueOnce(commits);
      getCommitsForPR.mockReturnValueOnce([
        {
          author: {
            login: 'andrew'
          }
        }
      ]);
      getUserByUsername.mockReturnValueOnce({
        login: 'andrew',
        name: 'Andrew Lisowski'
      });
      getUserByEmail.mockReturnValueOnce({
        login: 'adam',
        name: 'Adam Dierkens'
      });
      getUserByEmail.mockReturnValueOnce({
        login: 'adam',
        name: 'Adam Dierkens'
      });

      const gh = new GitHubRelease({ owner: 'Adam Dierkens', repo: 'test' });
      const modifiedCommits = await gh.getCommits('12345', '1234');
      expect(getUserByUsername).toHaveBeenCalled();
      expect(modifiedCommits).toMatchSnapshot();
    });
  });

  test('publish', async () => {
    const gh = new GitHubRelease();
    await gh.publish('release notes', '1.2.3');
    expect(publish).toHaveBeenCalled();
  });

  test('getLabels', async () => {
    const gh = new GitHubRelease();
    await gh.getLabels(123);
    expect(getLabels).toHaveBeenCalled();
  });

  test('getLatestRelease', async () => {
    const gh = new GitHubRelease();
    await gh.getLatestRelease();
    expect(getLatestRelease).toHaveBeenCalled();
  });

  test('getLatestRelease', async () => {
    const gh = new GitHubRelease();
    await gh.getPullRequest(22);
    expect(getLatestRelease).toHaveBeenCalled();
  });

  test('getSha', async () => {
    const gh = new GitHubRelease();
    await gh.getSha();
    expect(getLatestRelease).toHaveBeenCalled();
  });

  test('createStatus', async () => {
    const gh = new GitHubRelease();
    await gh.createStatus({
      state: 'pending',
      sha: '',
      number: 22,
      context: 'foo',
      url: 'google.com',
      description: 'testing'
    });
    expect(getLatestRelease).toHaveBeenCalled();
  });

  test('createComment', async () => {
    const gh = new GitHubRelease();
    await gh.createComment('Some long message', 22);
    expect(createComment).toHaveBeenCalled();
  });

  test('getPullRequests', async () => {
    const gh = new GitHubRelease();
    await gh.getPullRequests({ state: 'closed' });
    expect(getPullRequests).toHaveBeenCalled();
  });

  describe('addToChangelog', async () => {
    test("creates new changelog if one didn't exist", async () => {
      const gh = new GitHubRelease();
      await gh.addToChangelog(
        '# My new Notes',
        'klajsdlfk4lj51l43k5hj234l',
        'v0.0.0'
      );

      expect(writeSpy.mock.calls[0][1].includes(`# My new Notes`)).toBe(true);
    });

    test("creates new changelog if one didn't exist", async () => {
      const gh = new GitHubRelease();
      await gh.addToChangelog('# My new Notes', 'v1.0.0', 'v1.0.0', false);

      expect(writeSpy.mock.calls[0][1].includes(`v1.0.1`)).toBe(true);
    });

    test('prepends to old changelog', async () => {
      const gh = new GitHubRelease();

      existsSync.mockReturnValueOnce(true);
      readResult = '# My old Notes';

      await gh.addToChangelog(
        '# My new Notes',
        'asdfasdlkfjlkj435l2j',
        'v0.0.0'
      );
      expect(writeSpy.mock.calls[0][1].includes(readResult)).toBe(true);
    });

    test('should be able to configure message', async () => {
      const gh = new GitHubRelease();
      const message = 'pony foo';

      existsSync.mockReturnValueOnce(true);
      readResult = '# My old Notes';

      await gh.addToChangelog(
        '# My new Notes',
        'asdklfhlkh24387513',
        'v0.0.0',
        false,
        message
      );
      expect(execSpy.mock.calls[1][0].includes(message)).toBe(true);
    });
  });

  test('postToSlack', async () => {
    const gh = new GitHubRelease(undefined, {
      logger: dummyLog(),
      slack: 'https://custom-slack-url'
    });
    await gh.postToSlack('# My Notes', 'v1.0.0');
    expect(slackSpy).toHaveBeenCalled();
  });

  describe('generateReleaseNotes', async () => {
    test('should default to HEAD', async () => {
      const gh = new GitHubRelease();
      expect(await gh.generateReleaseNotes('1234')).toBe('');
    });

    test('should use configured HEAD', async () => {
      const gh = new GitHubRelease();
      expect(await gh.generateReleaseNotes('1234', '123')).toBe('');
    });

    test('should allow user to configure section headings', async () => {
      const gh = new GitHubRelease();

      const commits = [
        makeCommitFromMsg('First (#1234)'),
        makeCommitFromMsg('Second (#1235)'),
        makeCommitFromMsg('Third (#1236)'),
        makeCommitFromMsg('Fourth (#1237)')
      ];

      getGitLog.mockReturnValueOnce(commits);
      getLabels.mockReturnValueOnce(['major']);
      getLabels.mockReturnValueOnce(['minor']);
      getLabels.mockReturnValueOnce(['documentation', 'internal']);
      getLabels.mockReturnValueOnce(['patch']);

      expect(await gh.generateReleaseNotes('1234', '123')).toMatchSnapshot();
    });
  });

  describe('getSemverBump', () => {
    test('default to patch', async () => {
      const gh = new GitHubRelease();
      const commits = [
        makeCommitFromMsg('First'),
        makeCommitFromMsg('Second'),
        makeCommitFromMsg('Third')
      ];

      getGitLog.mockReturnValueOnce(commits);

      expect(await gh.getSemverBump('1234')).toBe(SEMVER.patch);
    });

    test('should use higher version', async () => {
      const gh = new GitHubRelease();
      const commits = [
        makeCommitFromMsg('First (#1234)'),
        makeCommitFromMsg('Second'),
        makeCommitFromMsg('Third')
      ];

      getGitLog.mockReturnValueOnce(commits);
      getLabels.mockReturnValueOnce(['minor']);

      expect(await gh.getSemverBump('1234', '123')).toBe(SEMVER.minor);
    });

    test('should not publish a release', async () => {
      const gh = new GitHubRelease();
      const commits = [
        makeCommitFromMsg('First (#1234)'),
        makeCommitFromMsg('Second (#1235)'),
        makeCommitFromMsg('Third (#1236)')
      ];

      getGitLog.mockReturnValueOnce(commits);
      getLabels.mockReturnValueOnce(['skip-release', 'patch']);
      getLabels.mockReturnValueOnce(['skip-release', 'patch']);
      getLabels.mockReturnValueOnce(['skip-release', 'minor']);

      expect(await gh.getSemverBump('1234', '123')).toBe('');
    });

    test('should publish a release', async () => {
      const gh = new GitHubRelease();
      const commits = [
        makeCommitFromMsg('First (#1234)'),
        makeCommitFromMsg('Second (#1235)'),
        makeCommitFromMsg('Third (#1236)')
      ];

      getGitLog.mockReturnValueOnce(commits);
      getLabels.mockReturnValueOnce(['patch']);
      getLabels.mockReturnValueOnce(['skip-release', 'patch']);
      getLabels.mockReturnValueOnce(['skip-release', 'minor']);

      expect(await gh.getSemverBump('1234', '123')).toBe(SEMVER.minor);
    });

    test('should default to publish a prepatch', async () => {
      const gh = new GitHubRelease();
      const commits = [
        makeCommitFromMsg('First (#1234)'),
        makeCommitFromMsg('Second (#1235)'),
        makeCommitFromMsg('Third (#1236)')
      ];

      getGitLog.mockReturnValueOnce(commits);
      getLabels.mockReturnValueOnce(['skip-release']);
      getLabels.mockReturnValueOnce([]);
      getLabels.mockReturnValueOnce([]);

      expect(await gh.getSemverBump('1234', '123')).toBe('');
    });

    test('should not publish a release in onlyPublishWithReleaseLabel without label', async () => {
      const gh = new GitHubRelease();
      const commits = [
        makeCommitFromMsg('First (#1234)'),
        makeCommitFromMsg('Second (#1235)'),
        makeCommitFromMsg('Third (#1236)')
      ];

      getGitLog.mockReturnValueOnce(commits);
      getLabels.mockReturnValueOnce(['patch']);
      getLabels.mockReturnValueOnce(['major']);
      getLabels.mockReturnValueOnce(['patch']);

      expect(await gh.getSemverBump('1234', '123', true)).toBe('');
    });

    test('should publish a release in onlyPublishWithReleaseLabel with label', async () => {
      const gh = new GitHubRelease();
      const commits = [
        makeCommitFromMsg('First (#1234)'),
        makeCommitFromMsg('Second (#1235)'),
        makeCommitFromMsg('Third (#1236)')
      ];

      getGitLog.mockReturnValueOnce(commits);
      getLabels.mockReturnValueOnce(['release', 'patch']);
      getLabels.mockReturnValueOnce(['patch']);
      getLabels.mockReturnValueOnce(['minor']);

      expect(await gh.getSemverBump('1234', '123', true)).toBe(SEMVER.minor);
    });

    test('should be able to configure labels', async () => {
      const labels = new Map();
      labels.set(SEMVER.major, 'Version: Major');
      labels.set(SEMVER.minor, 'Version: Minor');
      labels.set(SEMVER.patch, 'Version: Patch');
      labels.set('release', 'Deploy');

      const gh = new GitHubRelease(undefined, {
        logger,
        labels
      });
      const commits = [
        makeCommitFromMsg('First (#1234)'),
        makeCommitFromMsg('Second (#1235)'),
        makeCommitFromMsg('Third (#1236)')
      ];

      // Test default labels do nothing
      getGitLog.mockReturnValueOnce(commits);
      getLabels.mockReturnValueOnce(['Version: Major']);
      getLabels.mockReturnValueOnce(['Version: Patch']);
      getLabels.mockReturnValueOnce(['Version: Minor', 'release']);

      expect(await gh.getSemverBump('1234', '123', true)).toBe('');

      getGitLog.mockReturnValueOnce(commits);
      getLabels.mockReturnValueOnce(['Version: Minor', 'Deploy']);
      getLabels.mockReturnValueOnce(['Version: Major']);
      getLabels.mockReturnValueOnce(['Version: Patch']);

      expect(await gh.getSemverBump('1234', '123', true)).toBe(SEMVER.major);
    });
  });

  describe('addLabelsToProject', () => {
    test('should add labels', async () => {
      const gh = new GitHubRelease();
      const labels = new Map<VersionLabel, string>();
      labels.set(SEMVER.major, '1');
      labels.set(SEMVER.minor, '2');
      labels.set(SEMVER.patch, '3');

      await gh.addLabelsToProject(labels);

      expect(createLabel).toHaveBeenCalledWith(SEMVER.major, '1');
      expect(createLabel).toHaveBeenCalledWith(SEMVER.minor, '2');
      expect(createLabel).toHaveBeenCalledWith(SEMVER.patch, '3');
    });

    test('should not add old labels', async () => {
      const gh = new GitHubRelease();
      const labels = new Map<VersionLabel, string>();
      labels.set(SEMVER.major, '1');
      labels.set(SEMVER.minor, '2');

      getProjectLabels.mockReturnValueOnce(['1']);
      await gh.addLabelsToProject(labels);

      expect(createLabel).not.toHaveBeenCalledWith(SEMVER.major, '1');
      expect(createLabel).toHaveBeenCalledWith(SEMVER.minor, '2');
    });

    test('should add release label in onlyPublishWithReleaseLabel mode', async () => {
      const gh = new GitHubRelease();
      const labels = new Map<VersionLabel, string>();
      labels.set('release', 'deploy');

      await gh.addLabelsToProject(labels);
      expect(createLabel).not.toHaveBeenCalledWith('release', 'deploy');

      await gh.addLabelsToProject(labels, true);
      expect(createLabel).toHaveBeenCalledWith('release', 'deploy');
    });

    test('should add skip-release label not in onlyPublishWithReleaseLabel mode', async () => {
      const gh = new GitHubRelease();
      const labels = new Map<VersionLabel, string>();
      labels.set('skip-release', 'no!');

      await gh.addLabelsToProject(labels, true);
      expect(createLabel).not.toHaveBeenCalledWith('skip-release', 'no!');

      await gh.addLabelsToProject(labels);
      expect(createLabel).toHaveBeenCalledWith('skip-release', 'no!');
    });
  });
});
