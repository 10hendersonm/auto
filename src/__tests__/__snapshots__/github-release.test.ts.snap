// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GitHubRelease generateReleaseNotes should allow user to configure section headings 1`] = `
"#### 💥  Breaking Change

- First [#1234](https://github.com/web/site/pull/1234) (adam@dierkens.com)

#### 🚀  Enhancement

- Second [#1235](https://github.com/web/site/pull/1235) (adam@dierkens.com)

#### 🐛  Bug Fix

- Fourth [#1237](https://github.com/web/site/pull/1237) (adam@dierkens.com)

#### 🏠  Internal

- Third [#1236](https://github.com/web/site/pull/1236) (adam@dierkens.com)

#### Authors: 1

- Adam Dierkens (adam@dierkens.com)"
`;

exports[`GitHubRelease getCommits should resolve authors with PR commits 1`] = `
Array [
  Object {
    "authorEmail": "adam@dierkens.com",
    "authorName": "Adam Dierkens",
    "authors": Array [
      Object {
        "login": "adam",
        "name": "Adam Dierkens",
        "username": "adam",
      },
    ],
    "hash": "foo",
    "labels": Array [],
    "packages": undefined,
    "subject": "First",
  },
  Object {
    "authorEmail": "lisowski54@gmail.com",
    "authorName": "Andrew Lisowski",
    "authors": Array [
      Object {
        "login": "andrew",
        "name": "Andrew Lisowski",
        "username": "andrew",
      },
    ],
    "hash": "foo",
    "labels": undefined,
    "packages": undefined,
    "pullRequest": Object {
      "number": 123,
    },
    "subject": "Second",
  },
  Object {
    "authorEmail": "adam@dierkens.com",
    "authorName": "Adam Dierkens",
    "authors": Array [
      Object {
        "login": "adam",
        "name": "Adam Dierkens",
        "username": "adam",
      },
    ],
    "hash": "foo",
    "labels": Array [],
    "packages": undefined,
    "subject": "Third",
  },
]
`;
